src/util/valuetransformer.cpp

```diff
@@ -1,35 +1,124 @@
 #include "util/valuetransformer.h"
 
-ValueTransformer::ValueTransformer() {
-}
+#include <QDomElement>
+#include <QDomNode>
+#include <QDomNodeList>
+#include <memory>
+
+#include "skin/legacy/skincontext.h"
+
+class TransformNode {
+  public:
+    TransformNode() {
+    }
+    virtual ~TransformNode() {
+    }
 
-void ValueTransformer::addTransformer(TransformNode* pTransformer) {
-    m_transformers.append(pTransformer);
+    virtual double transform(double argument) const = 0;
+    virtual double transformInverse(double argument) const = 0;
+};
+
+namespace {
+
+class TransformAdd : public TransformNode {
+  public:
+    TransformAdd(double addend)
+            : m_addend(addend) {
+    }
+
+    double transform(double argument) const override {
+        return argument + m_addend;
+    }
+
+    double transformInverse(double argument) const override {
+        return argument - m_addend;
+    }
+
+  private:
+    double m_addend;
+};
+
+class TransformInvert : public TransformNode {
+  public:
+    double transform(double argument) const override {
+        return -argument;
+    }
+
+    double transformInverse(double argument) const override {
+        return -argument;
+    }
+};
+
+class TransformNot : public TransformNode {
+  public:
+    double transform(double argument) const override {
+        return !static_cast<bool>(argument);
+    }
+
+    double transformInverse(double argument) const override {
+        return !static_cast<bool>(argument);
+    }
+};
+
+class TransformIsEqual : public TransformNode {
+  public:
+    TransformIsEqual(double compareValue)
+            : m_compareValue(compareValue) {
+    }
+
+    double transform(double argument) const override {
+        return argument == m_compareValue;
+    }
+
+    double transformInverse(double argument) const override {
+        if (argument > 0.0) {
+            return m_compareValue;
```

Comment:
Is that correct? Maybe we should add a brief comment.

Reply:
didn't bother checking, I just moved this part from the header into the cpp.